src/blkdev.c

```diff
@@ -0,0 +1,384 @@
+}
+
+static int block_configure(block_t *block, const char **err)
+{
+  int rc;
+
+  if (uk_blkdev_state_get(block->dev) != UK_BLKDEV_UNCONFIGURED) {
+    *err = "Block device not in unconfigured state";
+    return -1;
+  }
+
+  struct uk_blkdev_info info;
+  rc = uk_blkdev_get_info(block->dev, &info);
+  if (rc) {
+    *err = "Error getting device information";
+    return -1;
+  }
+
+  if (QUEUE_COUNT > info.max_queues) {
+    *err = "Number of queue not supported";
+    return -1;
+  }
+
+  struct uk_blkdev_conf conf = { 0 };
+  conf.nb_queues = QUEUE_COUNT;
+  rc = uk_blkdev_configure(block->dev, &conf);
+  if (rc) {
+    *err = "Error configuring device";
+    return -1;
+  }
+
+  struct uk_blkdev_queue_info q_info;
+  rc = uk_blkdev_queue_get_info(block->dev, 0, &q_info);
+  if (rc) {
+    *err = "Error getting device queue information";
+    return -1;
+  }
+
+  assert(MAX_BLK_TOKENS < q_info.nb_max);
+
+  struct uk_blkdev_queue_conf q_conf = { 0 };
+  q_conf.a = block->alloc;
+  q_conf.callback = queue_callback;
+  q_conf.callback_cookie = block;
+  q_conf.s = uk_sched_current();
+  // nb_desc = 0: device default queue length
+  rc = uk_blkdev_queue_configure(block->dev, 0, 0, &q_conf);
+  if (rc) {
+    *err = "Error configuring device queue";
+    return -1;
+  }
+
+  rc = uk_blkdev_start(block->dev);
+  if (rc) {
+    *err = "Error starting block device";
+    return -1;
+  }
+
+  if (uk_blkdev_state_get(block->dev) != UK_BLKDEV_RUNNING) {
+    *err = "Block device not in running state";
+    return -1;
+  }
+
+  const struct uk_blkdev_cap* cap = uk_blkdev_capabilities(block->dev);
+  if (cap->mode == O_WRONLY) {
+    *err = "Device in write-only mode are not supported";
+    return -1;
+  }
+
+  rc = uk_blkdev_queue_intr_enable(block->dev, 0);
+  if (rc) {
+    *err = "Error enalbing interrupt for block device queue";
+    return -1;
+  }
+  return 0;
+}
+
+// -------------------------------------------------------------------------- //
+
+#define CAML_NAME_SPACE
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/alloc.h>
+#include <caml/bigarray.h>
+
+value uk_block_init(value v_id)
+{
+  CAMLparam1(v_id);
+  CAMLlocal2(v_result, v_error);
+  const char *err = NULL;
+  const unsigned id = Int_val(v_id);
+
+  UK_ASSERT(id >= 0);
+
+  block_t *block = block_get(id, &err);
+  if (!block) {
+    v_result = alloc_result_error(err);
+    CAMLreturn(v_result);
+  }
+
+  const int rc = block_configure(block, &err);
+  if (rc) {
+    v_result = alloc_result_error(err);
+    CAMLreturn(v_result);
+  }
+
+  v_result = alloc_result_ok(Val_ptr(block));
+  CAMLreturn(v_result);
+}
+
+value uk_block_info(value v_block)
+{
+  CAMLparam1(v_block);
+  CAMLlocal1(v_result);
+
+  block_t *block = (block_t*)Ptr_val(v_block);
+  const struct uk_blkdev_cap* cap = uk_blkdev_capabilities(block->dev);
+
+  v_result = caml_alloc(3, 0);
+  // read_write
+  Store_field(v_result, 0, (cap->mode == O_RDONLY) ? Val_false : Val_true);
```

Comment:
I think it’s usually presented as follows:
```suggestion
  Store_field(v_result, 0, (cap->mode == O_RDONLY) ? Val_false : Val_true);
                                                           // read_write
```